src/vs/workbench/contrib/terminal/common/terminalEnvironment.ts

```diff
@@ -105,77 +105,80 @@ export function shouldSetLangEnvVariable(env: IProcessEnvironment, detectLocale:
-	const n = parts.length;
-	if (n === 0) {
+	if (!locale) {
 		// Fallback to en_US if the locale is unknown
 		return 'en_US.UTF-8';
 	}
-	if (n === 1) {
-		// The local may only contain the language, not the variant, if this is the case guess the
-		// variant such that it can be used as a valid $LANG variable. The language variant chosen
-		// is the original and/or most prominent with help from
-		// https://stackoverflow.com/a/2502675/1156119
-		// The list of locales was generated by running `locale -a` on macOS
-		const languageVariants: { [key: string]: string } = {
-			af: 'ZA',
-			am: 'ET',
-			be: 'BY',
-			bg: 'BG',
-			ca: 'ES',
-			cs: 'CZ',
-			da: 'DK',
-			// de: 'AT',
-			// de: 'CH',
-			de: 'DE',
-			el: 'GR',
-			// en: 'AU',
-			// en: 'CA',
-			// en: 'GB',
-			// en: 'IE',
-			// en: 'NZ',
-			en: 'US',
-			es: 'ES',
-			et: 'EE',
-			eu: 'ES',
-			fi: 'FI',
-			// fr: 'BE',
-			// fr: 'CA',
-			// fr: 'CH',
-			fr: 'FR',
-			he: 'IL',
-			hr: 'HR',
-			hu: 'HU',
-			hy: 'AM',
-			is: 'IS',
-			// it: 'CH',
-			it: 'IT',
-			ja: 'JP',
-			kk: 'KZ',
-			ko: 'KR',
-			lt: 'LT',
-			// nl: 'BE',
-			nl: 'NL',
-			no: 'NO',
-			pl: 'PL',
-			pt: 'BR',
-			// pt: 'PT',
-			ro: 'RO',
-			ru: 'RU',
-			sk: 'SK',
-			sl: 'SI',
-			sr: 'YU',
-			sv: 'SE',
-			tr: 'TR',
-			uk: 'UA',
-			zh: 'CN',
-		};
-		if (parts[0] in languageVariants) {
-			parts.push(languageVariants[parts[0]]);
-		}
-	} else {
-		// Ensure the variant is uppercase to be a valid $LANG
-		parts[1] = parts[1].toUpperCase();
+
+	// Special handling for zh-hans and zh-hant languages
+	if (locale.startsWith('zh-hans')) {
+		return 'zh_CN.UTF-8';
+	} else if (locale.startsWith('zh-hant')) {
+		return 'zh_TW.UTF-8';
+	}
+
+	const parts = locale.split('-');
+
+	// The locale may only contain the language, not the variant, if this is the case guess the
+	// variant such that it can be used as a valid $LANG variable. The language variant chosen
+	// is the original and/or most prominent with help from
+	// https://stackoverflow.com/a/2502675/1156119
+	// The list of locales was generated by running `locale -a` on macOS
+	const languageVariants: { [key: string]: string } = {
+		af: 'ZA',
+		am: 'ET',
+		be: 'BY',
+		bg: 'BG',
+		ca: 'ES',
+		cs: 'CZ',
+		da: 'DK',
+		// de: 'AT',
+		// de: 'CH',
+		de: 'DE',
+		el: 'GR',
+		// en: 'AU',
+		// en: 'CA',
+		// en: 'GB',
+		// en: 'IE',
+		// en: 'NZ',
+		en: 'US',
+		es: 'ES',
+		et: 'EE',
+		eu: 'ES',
+		fi: 'FI',
+		// fr: 'BE',
+		// fr: 'CA',
+		// fr: 'CH',
+		fr: 'FR',
+		he: 'IL',
+		hr: 'HR',
+		hu: 'HU',
+		hy: 'AM',
+		is: 'IS',
+		// it: 'CH',
+		it: 'IT',
+		ja: 'JP',
+		kk: 'KZ',
+		ko: 'KR',
+		lt: 'LT',
+		// nl: 'BE',
+		nl: 'NL',
+		no: 'NO',
+		pl: 'PL',
+		pt: 'BR',
+		// pt: 'PT',
+		ro: 'RO',
+		ru: 'RU',
+		sk: 'SK',
+		sl: 'SI',
+		sr: 'YU',
+		sv: 'SE',
+		tr: 'TR',
+		uk: 'UA',
+		zh: 'CN',
+	};
+	if (parts[0] in languageVariants) {
+		return `${parts[0]}_${languageVariants[parts[0]]}.UTF-8`;
 	}
```

Comment:
I think we still need to `n === 2` case, can we keep this but transform the new `zh-hans` and `zh-hant` cases before splitting the string? `if (==`zh-hans` || == `zh-hant`) { locale = 'zh-CN'; }`

Reply:
`zh-hans` and `zh-hant` are handled before splitting the locale string. As for the other locales, `app.getLocale()` was returning just a single part anyway, so I'm not sure whether the `n === 2` case is valid still.